plugins/hls-qualify-imported-names-plugin/src/Ide/Plugin/QualifyImportedNames.hs

```diff
@@ -0,0 +1,259 @@
+                                                    PluginId,
+                                                    PluginMethodHandler,
+                                                    defaultPluginDescriptor,
+                                                    mkPluginHandler)
+import           Language.LSP.Types                (CodeAction (CodeAction, _command, _diagnostics, _disabled, _edit, _isPreferred, _kind, _title, _xdata),
+                                                    CodeActionKind (CodeActionQuickFix),
+                                                    CodeActionParams (CodeActionParams),
+                                                    Method (TextDocumentCodeAction),
+                                                    SMethod (STextDocumentCodeAction),
+                                                    TextDocumentIdentifier (TextDocumentIdentifier),
+                                                    TextEdit (TextEdit),
+                                                    WorkspaceEdit (WorkspaceEdit, _changeAnnotations, _changes, _documentChanges),
+                                                    type (|?) (InR),
+                                                    uriToNormalizedFilePath)
+
+thenCmp :: Ordering -> Ordering -> Ordering
+{-# INLINE thenCmp #-}
+thenCmp EQ       ordering = ordering
+thenCmp ordering _        = ordering
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor pluginId = (defaultPluginDescriptor pluginId) {
+  pluginHandlers = mconcat
+    [ mkPluginHandler STextDocumentCodeAction codeActionProvider
+    ]
+}
+
+isRangeWithinSrcSpan :: Range -> SrcSpan -> Bool
+isRangeWithinSrcSpan (Range start end) srcSpan =
+  isInsideSrcSpan start srcSpan && isInsideSrcSpan end srcSpan
+
+findLImportDeclAt :: Range -> ParsedModule -> Maybe (LImportDecl GhcPs)
+findLImportDeclAt range parsedModule
+  | ParsedModule {..} <- parsedModule
+  , L _ hsModule <- pm_parsed_source
+  , locatedImportDecls <- hsmodImports hsModule =
+      find (\ (L srcSpan _) -> isRangeWithinSrcSpan range srcSpan) locatedImportDecls
+  | otherwise = Nothing
+
+makeCodeActions :: Uri -> [TextEdit] -> [a |? CodeAction]
+makeCodeActions uri textEdits = [InR CodeAction {..} | not (null textEdits)]
+  where _title = "Qualify imported names"
+        _kind = Just CodeActionQuickFix
+        _command = Nothing
+        _edit = Just WorkspaceEdit {..}
+        _changes = Just $ HashMap.singleton uri $ List textEdits
+        _documentChanges = Nothing
+        _diagnostics = Nothing
+        _isPreferred = Nothing
+        _disabled = Nothing
+        _xdata = Nothing
+        _changeAnnotations = Nothing
+
+getTypeCheckedModule :: IdeState -> NormalizedFilePath -> IO (Maybe TcModuleResult)
+getTypeCheckedModule ideState normalizedFilePath =
+  runAction "QualifyImportedNames.TypeCheck" ideState (use TypeCheck normalizedFilePath)
```

Comment:
You may want to use `useWithStale` here, so that your plugin is able to work even when the file doesn't typecheck.

Reply:
I'm not sure that's good because if someone is typing names in a file that doesn't typecheck and expects those names to be qualified by the plugin then they might be confused as to why older used names are qualified, but the ones they've been typing aren't. I feel like the plugin is used like a tidy up or refactoring tool after the program generally works, and so it's okay to have the file typecheck before using the plugin.